common/flatpak-transaction-log.c

```diff
@@ -0,0 +1,305 @@
+      if (!g_error_matches (my_error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+        {
+          g_propagate_error (error, g_steal_pointer (&my_error));
+          return NULL;
+        }
+
+      toplevel = build_empty_array ();
+    }
+  
+  return toplevel;
+}
+
+static gboolean
+write_node_to_log_file (GFile        *file,
+                        JsonNode     *node,
+                        GCancellable *cancellable,
+                        GError       **error)
+{
+  g_autoptr(GFile) parent = g_file_get_parent (file);
+  g_autoptr(GFileOutputStream) stream = NULL;
+  g_autoptr(JsonGenerator) generator = NULL;
+  g_autoptr(GError) my_error = NULL;
+
+  if (!g_file_make_directory_with_parents (parent, cancellable, &my_error))
+    {
+      if (!g_error_matches (my_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        {
+          g_propagate_error (error, g_steal_pointer (&my_error));
+          return FALSE;
+        }
+    }
+
+  stream = g_file_replace (file,
+                           NULL,
+                           TRUE,
+                           G_FILE_CREATE_NONE,
+                           cancellable,
+                           error);
+
+  if (!stream)
+    return FALSE;
+
+  generator = json_generator_new ();
+  json_generator_set_root (generator, node);
+  json_generator_set_pretty (generator, TRUE);
+
+  if (!json_generator_to_stream (generator, G_OUTPUT_STREAM (stream), cancellable, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+#define EVENT_LOG_VERSION "1"
+
+static JsonNode *
+build_base_event (const gchar *event)
+{
+  g_autoptr(JsonBuilder) builder = json_builder_new ();
+
+  json_builder_begin_object (builder);
+
+  json_builder_set_member_name (builder, "event");
+  json_builder_add_string_value (builder, event);
+
+  json_builder_set_member_name (builder, "version");
+  json_builder_add_string_value (builder, EVENT_LOG_VERSION);
+  json_builder_end_object (builder);
+
+  return json_builder_get_root (builder);
+}
+
+static JsonNode *
+deploy_install_event (const gchar *ref,
+                      const gchar *origin)
+{
+  JsonNode *base = build_base_event ("deploy-install");
+  JsonObject *object = json_node_get_object (base);
+
+  json_object_set_string_member (object, "ref", ref);
+  json_object_set_string_member (object, "origin", origin);
+
+  return base;
+}
+
+gboolean
+flatpak_transaction_log_commit_deploy_install_event (FlatpakTransactionLog *log,
+		                                     const gchar           *ref,
+		                                     const gchar           *origin,
+		                                     GCancellable          *cancellable,
+		                                     GError                **error)
+{
+  g_autoptr(JsonNode) node = load_existing_log_if_exists (log->path, cancellable, error);
+  JsonArray *array = json_node_get_array (node);
+  g_autoptr(JsonNode) event = deploy_install_event (ref, origin);
+
+  json_array_add_object_element (array, json_node_dup_object (event));
```

Comment:
The log file right now is effectively unbounded. We could limit it to N entries, but the tradeoff is that the log becomes lossy and not as useful to consumers (and especially in our usecase where we want to know if a flatpak was ever uninstalled at some point, that information will be lost at some point, so we end up auto-installing a flatpak that was explicitly uninstalled before).